### PR TITLE
Improved abstractions + better object construction

### PR DESCRIPTION
**Pull request by truthbk**
Introduces breaking changes. 

Better description pending...

### CODE REVIEW

File: src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
+package com.timgroup.statsd;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.DatagramChannel;
+import java.nio.charset.Charset;
+import java.util.concurrent.Callable;
+
+import java.util.Queue;
```
Comment (ogaca-dd) [inline src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java]: Note: Several imports are unused.

File: src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        boolean done = false;
+        while(!done) {
+            try {
+                endSignal.await();
+                done = true;
+            } catch (final InterruptedException e) { }
+        }
+    }
+
+    boolean isShutdown() {
```
Comment (ogaca-dd) [inline src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java]: Note: As `shutdown` is defined in `StatsDProcessor`, `isShutdown()` and `shutdown()` could be moved to `StatsDProcessor`. Same comment could also be applied for `StatsDNonBlockingProcessor`.

File: src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Override
+    boolean send(final String message) {
+        if (!shutdown) {
+            if (qSize.get() < qCapacity) {
+                messages.offer(message);
+                qSize.incrementAndGet();
+                return true;
+            }
+        }
```
Comment (ogaca-dd) [inline src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java]: Would it make sense to catch `InterruptedException` as in https://github.com/DataDog/java-dogstatsd-client/pull/96/files#diff-d87f261783303f309bda2086a2e14756R35?

File: src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                                }
+                                sendBuffer.put(data);
+                                if (null == messages.peek()) {
+                                    outboundQueue.put(sendBuffer);
+                                    sendBuffer = bufferPool.borrow();
+                                }
+                            }
+                        } catch (final InterruptedException e) {
+                            if (shutdown) {
+                                return;
```
Comment (ogaca-dd) [inline src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java]: Do we need `endSignal.countDown();` as for https://github.com/DataDog/java-dogstatsd-client/pull/96/files#diff-d87f261783303f309bda2086a2e14756R86

File: src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java (hunk 1 above)
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java]: We have to add linting to the CI - thank you for checking. I don't use an IDE, so..... 😊

File: src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java (hunk 4 above)
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java]: nice catch, yup!

File: src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java (hunk 3 above)
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDNonBlockingProcessor.java]: I wouldn't expect an `InterruptedException` to happen there as we are dealing with a non-blocking queue in this case and the `offer()` method would just return as opposed to stay there waiting for the write.

File: src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java (hunk 2 above)
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDBlockingProcessor.java]: Yup, code was already defined in `StatsdProcessor` just forgot doing this cleanup. Thank you! Great catch again.

File: src/test/java/com/timgroup/statsd/NonBlockingStatsDClientMaxPerfTest.java [hunk 5]
```diff
@@ -41,43 +41,67 @@
...
-                 { 30, 17255, 2048, 1, 2 },  // 30 seconds, 17255 port, 2048 qSize, 1 sender worke, 2 processor workers
-                 { 30, 17255, 4096, 1, 2 },  // 30 seconds, 17255 port, 4096 qSize, 1 sender worke, 2 processor workers
-                 // // { 30, 17255, Integer.MAX_VALUE, 1, 2 },  // 30 seconds, 17255 port, MAX_VALUE qSize, 1 worker
-                 { 30, 17255, 256, 2, 2 },  // 30 seconds, 17255 port, 256 qSize, 2 sender workers, 2 processor workers
-                 { 30, 17255, 512, 2, 2 },  // 30 seconds, 17255 port, 512 qSize, 2 sender workers, 2 processor workers
-                 { 30, 17255, 1024, 2, 2 },  // 30 seconds, 17255 port, 1024 qSize, 2 sender workers, 2 processor workers
-                 { 30, 17255, 2048, 2, 2 },  // 30 seconds, 17255 port, 2048 qSize, 2 sender workers, 2 processor workers
-                 { 30, 17255, 4096, 2, 2 }  // 30 seconds, 17255 port, 4096 qSize, 2 sender workers, 2 processor workers
-                 // // { 30, 17255, Integer.MAX_VALUE, 2 }  // 30 seconds, 17255 port, MAX_VALUE qSize, 2 sender workers
+                 { 30, false, 256, 1, 1 },  // 30 seconds, non-blocking, 256 qSize, 1 worker
```
Comment (ogaca-dd) [inline src/test/java/com/timgroup/statsd/NonBlockingStatsDClientMaxPerfTest.java]: Note: This collection could be created with 3 "for loop" (non-blocking, Qsize, worker count)

**Comment by truthbk** [comment]
Addressed changes + fixed CI to skip `checkstyle` on Java7, and compile successfully. 

Merging onto #95

**Pull request merged**